analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.Enumerable.cs

```diff
@@ -69,25 +69,64 @@ internal sealed partial class Invocation
         "UnionBy",
         nameof(Enumerable.Where),
         nameof(Enumerable.Zip),
-    };
+    ];
+
+    private static readonly HashSet<string> ElemenetExistsCheckMethods =
+    [
+        nameof(Enumerable.Contains),
+        nameof(Enumerable.Any),
+        nameof(List<int>.Exists)
+    ];
 
     private static ProgramState[] ProcessLinqEnumerableAndQueryable(ProgramState state, IInvocationOperationWrapper invocation)
     {
         var name = invocation.TargetMethod.Name;
+        var states = ProcessBoolCollectionMethods(state, invocation);
         if (ReturningNotNull.Contains(name))
         {
-            return state.SetOperationConstraint(invocation, ObjectConstraint.NotNull).ToArray();
+            return states.Select(x => x.SetOperationConstraint(invocation, ObjectConstraint.NotNull)).ToArray();
         }
         // ElementAtOrDefault is intentionally not supported. It's causing many FPs
         else if (name is nameof(Enumerable.FirstOrDefault) or nameof(Enumerable.LastOrDefault) or nameof(Enumerable.SingleOrDefault))
         {
-            return invocation.TargetMethod.ReturnType.IsReferenceType
-                ? [state.SetOperationConstraint(invocation, ObjectConstraint.Null), state.SetOperationConstraint(invocation, ObjectConstraint.NotNull)]
-                : state.ToArray();
+            return states.SelectMany(x => new List<ProgramState>
+            {
+                x.SetOperationConstraint(invocation, ObjectConstraint.Null),
+                x.SetOperationConstraint(invocation, ObjectConstraint.NotNull)
+            }).ToArray();
         }
         else
         {
-            return state.ToArray();
+            return states;
+        }
+    }
+
+    private static ProgramState[] ProcessBoolCollectionMethods(ProgramState state, IInvocationOperationWrapper invocation)
+    {
+        if (invocation.TryGetInstance(state) is { } instance && ElemenetExistsCheckMethods.Contains(invocation.TargetMethod.Name) && instance.TrackedSymbol(state) is { } instanceSymbol)
+        {
+            return state[instanceSymbol]?.Constraint<CollectionConstraint>() switch
+            {
+                CollectionConstraint constraint when constraint == CollectionConstraint.Empty => state.SetOperationConstraint(invocation, BoolConstraint.False).ToArray(),
+                CollectionConstraint constraint when constraint == CollectionConstraint.NotEmpty && HasNoParameters(invocation.TargetMethod) =>
+                    state.SetOperationConstraint(invocation, BoolConstraint.True).ToArray(),
+                CollectionConstraint constraint when constraint == CollectionConstraint.NotEmpty => state.ToArray(),
```

Comment:
Alternatively:
```suggestion
                CollectionConstraint constraint when constraint == CollectionConstraint.NotEmpty => HasNoParameters(invocation.TargetMethod)
                    ? state.SetOperationConstraint(invocation, BoolConstraint.True).ToArray()
                    : state.ToArray(),
```
Not sure, which version I prefer.

Reply:
I think I like the one you propose here but it might be visually more confusing. Not sure.